predicators/envs/sandwich.py

```diff
@@ -0,0 +1,832 @@
+                                self._Clear_holds)
+        self._BoardClear = Predicate("BoardClear", [self._board_type],
+                                     self._BoardClear_holds)
+        self._IsBread = Predicate("IsBread", [self._ingredient_type],
+                                  self._IsBread_holds)
+        self._IsBurger = Predicate("IsBurger", [self._ingredient_type],
+                                   self._IsBurger_holds)
+        self._IsHam = Predicate("IsHam", [self._ingredient_type],
+                                self._IsHam_holds)
+        self._IsEgg = Predicate("IsEgg", [self._ingredient_type],
+                                self._IsEgg_holds)
+        self._IsCheese = Predicate("IsCheese", [self._ingredient_type],
+                                   self._IsCheese_holds)
+        self._IsLettuce = Predicate("IsLettuce", [self._ingredient_type],
+                                    self._IsLettuce_holds)
+        self._IsTomato = Predicate("IsTomato", [self._ingredient_type],
+                                   self._IsTomato_holds)
+        self._IsGreenPepper = Predicate("IsGreenPepper",
+                                        [self._ingredient_type],
+                                        self._IsGreenPepper_holds)
+        # Options
+        self._Pick: ParameterizedOption = utils.SingletonParameterizedOption(
+            # variables: [robot, object to pick]
+            "Pick",
+            self._Pick_policy,
+            types=[self._robot_type, self._ingredient_type])
+        self._Stack: ParameterizedOption = utils.SingletonParameterizedOption(
+            # variables: [robot, object on which to stack currently-held-object]
+            "Stack",
+            self._Stack_policy,
+            types=[self._robot_type, self._ingredient_type])
+        self._PutOnBoard: ParameterizedOption = \
+            utils.SingletonParameterizedOption(
+            # variables: [robot, board]
+            "PutOnBoard",
+            self._PutOnBoard_policy,
+            types=[self._robot_type, self._board_type])
+        # Static objects (always exist no matter the settings).
+        self._robot = Object("robby", self._robot_type)
+        self._holder = Object("holder", self._holder_type)
+        self._board = Object("board", self._board_type)
+        self._num_ingredients_train = CFG.sandwich_ingredients_train
+        self._num_ingredients_test = CFG.sandwich_ingredients_test
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "sandwich"
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        x, y, z, fingers = action.arr
+        # Infer which transition function to follow
+        if fingers < self.held_tol:
+            return self._transition_pick(state, x, y, z)
+        thickness = self.ingredient_thickness + self.board_thickness
+        if z < self.table_height + thickness:
```

Comment:
I think I understand this, but could you leave a quick comment about why this logic is right here?